### PR TITLE
add AQ_NO_MODIFIERS to try to mitigate hardware limitations

### LINKED ISSUE #76: Blank screen with usb/thunderbolt dock or usb-c -> hdmi or igpu
basically AQ fails to modeset some limited hardware for reasons...
possibly related issues hyprwm/Hyprland#7321 #59 #40

### to test if you have this
get a [DRM log](https://wiki.hyprland.org/Crashes-and-Bugs/#debugging-drm-issues) and if it contains the below, then you have some hardware limitation
```
Requested display configuration exceeds system DDB limitations
```
btw @romanstingler made a script if you want to use it
https://github.com/romanstingler/scripts/blob/main/hyprland_drm_debug.sh

---

**known causes**
* limited graphics hardware, Ex. integrated graphics or old gpus trying to do 4k
* usb hub with limited bandwidth, Ex. it can't transmit 4k
* sometimes usb-c to hdmi cables. with limited bandwidth?
* thunderbolt dock with limited bandwidth

**it can show up in multiple ways**
* screen is on but not displaying anything
* in hyprland logs AQ spams `Invalid Argument` or `No format for output`
* in dmesg [DRM logs](https://wiki.hyprland.org/Crashes-and-Bugs/#debugging-drm-issues) repeating `Requested display configuration exceeds system DDB limitations`

**possible mitigations (untested)**
* lower resolution or refresh rate on monitors
* use a different hub or cable
* use a discrete gpu
* use less monitors
* play around with different monitor to plug configurations until it hopefully works

### PR DESCRIPTION
**Pull request by ikalco**
will hopefully mitigate or fix #76

plz test @romanstingler @techyporcupine

### CODE REVIEW

**Comment by romanstingler** [comment]
I can confirm that this indeed fixes the issues
all monitors are displaying the set resolution 
```
Monitor eDP-1 (ID 0):
	2560x1440@60.00000 at 3840x0
	description: California Institute of Technology 0x1500
	make: California Institute of Technology
	model: 0x1500
	serial: 
	active workspace: 1 (1)
	special workspace: 0 ()
	reserved: 0 35 0 0
	scale: 1.00
	transform: 0
	focused: no
	dpmsStatus: 1
	vrr: 0
	activelyTearing: false
	disabled: false
	currentFormat: A XRGB8888 H XRGB8888
	availableModes: 3840x2160@60.00Hz 

Monitor DP-3 (ID 1):
	4096x2160@59.94000 at 6400x0
	description: Hisense Electric Co. Ltd. HISENSE 0x676C626C
	make: Hisense Electric Co., Ltd.
	model: HISENSE
	serial: 0x676C626C
	active workspace: 2 (2)
	special workspace: 0 ()
	reserved: 0 35 0 0
	scale: 1.00
	transform: 0
	focused: yes
	dpmsStatus: 1
	vrr: 0
	activelyTearing: false
	disabled: false
	currentFormat: A XRGB8888 H XRGB8888
	availableModes: 3840x2160@60.00Hz 4096x2160@60.00Hz 4096x2160@59.94Hz 4096x2160@50.00Hz 4096x2160@30.00Hz 4096x2160@29.97Hz 4096x2160@25.00Hz 4096x2160@24.00Hz 4096x2160@23.98Hz 3840x2160@60.00Hz 3840x2160@60.00Hz 3840x2160@59.94Hz 3840x2160@59.94Hz 3840x2160@50.00Hz 3840x2160@50.00Hz 3840x2160@30.00Hz 3840x2160@29.97Hz 3840x2160@25.00Hz 3840x2160@24.00Hz 3840x2160@23.98Hz 2560x1440@59.95Hz 1920x1080@120.00Hz 1920x1080@119.88Hz 1920x1080@60.00Hz 1920x1080@59.94Hz 1920x1080@50.00Hz 1920x1080@30.00Hz 1920x1080@29.97Hz 1920x1080@25.00Hz 1920x1080@24.00Hz 1920x1080@23.98Hz 1280x1024@60.02Hz 1280x960@60.00Hz 1280x720@60.00Hz 1280x720@60.00Hz 1280x720@59.94Hz 1280x720@50.00Hz 1024x768@60.00Hz 800x600@60.32Hz 720x576@50.00Hz 720x480@60.00Hz 720x480@60.00Hz 720x480@59.94Hz 720x480@59.94Hz 640x480@60.00Hz 640x480@59.94Hz 

Monitor DP-1 (ID 2):
	3840x2160@60.00000 at 0x0
	description: Samsung Electric Company U32R59x 0x43573242
	make: Samsung Electric Company
	model: U32R59x
	serial: 0x43573242
	active workspace: 3 (3)
	special workspace: 0 ()
	reserved: 0 35 0 0
	scale: 1.00
	transform: 0
	focused: no
	dpmsStatus: 1
	vrr: 0
	activelyTearing: false
	disabled: false
	currentFormat: A XRGB8888 H XRGB8888
	availableModes: 3840x2160@60.00Hz 3840x2160@60.00Hz 3840x2160@59.94Hz 3840x2160@50.00Hz 3840x2160@30.00Hz 3840x2160@30.00Hz 3840x2160@29.97Hz 3840x2160@25.00Hz 3840x2160@24.00Hz 3840x2160@23.98Hz 2560x1440@59.95Hz 1920x1080@60.00Hz 1920x1080@60.00Hz 1920x1080@59.94Hz 1920x1080@50.00Hz 1920x1080@50.00Hz 1920x1080@30.00Hz 1920x1080@29.97Hz 1680x1050@59.95Hz 1600x900@60.00Hz 1280x1024@75.03Hz 1280x1024@60.02Hz 1440x900@59.89Hz 1280x800@59.81Hz 1152x864@75.00Hz 1280x720@60.00Hz 1280x720@60.00Hz 1280x720@59.94Hz 1280x720@50.00Hz 1024x768@75.03Hz 1024x768@70.07Hz 1024x768@60.00Hz 832x624@74.55Hz 800x600@75.00Hz 800x600@72.19Hz 800x600@60.32Hz 800x600@56.25Hz 720x576@50.00Hz 720x480@60.00Hz 720x480@60.00Hz 720x480@59.94Hz 720x480@59.94Hz 640x480@75.00Hz 640x480@72.81Hz 640x480@66.67Hz 640x480@60.00Hz 640x480@59.94Hz 720x400@70.08Hz 
```

**Comment by techyporcupine** [comment]
I'll try this out later today when I have my dock again!

File: src/backend/drm/DRM.cpp [hunk 1]
```diff
@@ -381,9 +381,15 @@ bool Aquamarine::CDRMBackend::checkFeatures() {
...
-    drmProps.supportsAsyncCommit     = drmGetCap(gpu->fd, DRM_CAP_ASYNC_PAGE_FLIP, &cap) == 0 && cap == 1;
-    drmProps.supportsAddFb2Modifiers = drmGetCap(gpu->fd, DRM_CAP_ADDFB2_MODIFIERS, &cap) == 0 && cap == 1;
-    drmProps.supportsTimelines       = drmGetCap(gpu->fd, DRM_CAP_SYNCOBJ_TIMELINE, &cap) == 0 && cap == 1;
+    drmProps.supportsAsyncCommit = drmGetCap(gpu->fd, DRM_CAP_ASYNC_PAGE_FLIP, &cap) == 0 && cap == 1;
+    drmProps.supportsTimelines   = drmGetCap(gpu->fd, DRM_CAP_SYNCOBJ_TIMELINE, &cap) == 0 && cap == 1;
+
+    if (envEnabled("AQ_NO_MODIFIERS")) {
+        backend->log(AQ_LOG_WARNING, "drm: AQ_NO_MODIFIERS enabled, disabling modifiers for DRM buffers.");
+        drmProps.supportsAddFb2Modifiers = false;
+    } else {
```
Comment (vaxerski) [inline src/backend/drm/DRM.cpp:390]: eeeeeeeeeeeeeeeee whats this brace

**Comment by techyporcupine** [comment]
How am I supposed to add the new branch into my flake? I know how to do this normally, but AQ is part of the hyprland flake so I'm not sure how to add it to test...

**Comment by techyporcupine** [comment]
Sorry, I'm unable to figure out how to specify env stuff on NixOS for building it, but if @romanstingler says it's working, I would guess it would work for me.

**Comment by fufexan** [comment]
@techyporcupine you can test with `AQ_NO_MODIFIERS=1 hyprland` in a tty.

**Comment by techyporcupine** [comment]
Hmm no luck.
Is this right in my flake?
```
    hyprland = {
      url = "git+https://github.com/hyprwm/Hyprland?submodules=1";
      inputs.aquamarine.url = "github:ikalco/aquamarine/add_aq_no_modifiers";
    };
```
I didn't notice it compiling aquamarine on a rebuild, is that a problem?

**Comment by ikalco** [comment]
you can verify if the fix is on by searching for `AQ_NO_MODIFIERS` in the logs, if it shows up its on

**Comment by techyporcupine** [comment]
Hmm, it's not showing up. Probably just keep going with the PR since one person reported it to fix it?

**Review by vaxerski — APPROVED** [review]

**Comment by techyporcupine** [comment]
Will this just automatically work now or do I need to pass any flags?

**Comment by romanstingler** [comment]
> Will this just automatically work now or do I need to pass any flags?

for now just add 
env = AQ_NO_MODIFIERS,1
to your `hyprland.conf`

no clue if in the future they will add the code like wlroots or if they will get the modifiers from something like 
`DRM_IOCTL_MODE_GETPLANE2` and set that on a per-plane basis.
https://lists.freedesktop.org/archives/intel-gfx/2017-January/116698.html

**Comment by romanstingler** [comment]
Vax I have a question, would you oppose a more detailed explanation of the env variables in the docs? 

something like 
```
AQ_NO_MODIFIERS: Disables DRM modifiers for buffers. 
This can also resolve issues where a monitor fails to turn on or is not detected properly.
DRM modifiers define how pixel data is stored in memory (e.g., in optimized formats). 
Setting AQ_NO_MODIFIERS to 1 ensures simpler buffer handling, 
which might be necessary on hardware that doesn’t support advanced memory layouts, 
especially for high resolutions like 4K.
```

```
AQ_DRM_DEVICES: This option allows you to set a specific list of DRM devices (GPUs) for the system to use. 
The list is separated by colons (:), and the first device listed will be the primary GPU. 
For example: /dev/dri/card1:/dev/dri/card0. 
This means the system will primarily use card1 and then fall back to card0 if needed.

AQ_NO_ATOMIC: Disables atomic mode setting for the DRM. 
Atomic mode setting is a feature that makes screen updates more reliable by grouping changes together. 
Setting AQ_NO_ATOMIC to 1 can be useful on hardware or drivers that don’t fully support atomic mode setting, 
as it can prevent crashes or freezes caused by incomplete or incorrect atomic updates.

AQ_MGPU_NO_EXPLICIT: Disables explicit synchronization between buffers when using multiple GPUs (multi-GPU setups).
Without explicit syncing, the system handles buffer transfers between GPUs more automatically, 
which can improve performance in some cases but might introduce errors in more complex setups.
```

**Comment by vaxerski** [comment]
fine by me

**Comment by techyporcupine** [comment]
Can confirm this fix works for me!

**Comment by techyporcupine** [comment]
This actually only seems to make to work in about 50% of my Hyprland sessions... any ideas?

**Pull request merged**